src/scheduler.rs

```diff
@@ -0,0 +1,1013 @@
+/// the existing mapping. This useful for when pods are deployed as containers on a standard K8S
+/// environment.
+struct HashingStrategy<'a> {
+    eligible_nodes: &'a RoleGroupEligibleNodes,
+    hasher: RefCell<DefaultHasher>,
+}
+
+pub enum ScheduleStrategy {
+    /// A scheduling strategy that will refuse to schedule two pods within one role+group on the same
+    /// node. If no enough pods are available, the pod will not be scheduled on any node.
+    /// This useful for when pods are deployed on a bare metal K8S environment with Stackable agents as nodes.
+    GroupAntiAffinity,
+    /// A scheduling strategy that will simply hash the pod onto one of the existing pods without
+    /// any consideration for the distribution of all other existing pods.This useful for when pods
+    /// are deployed as containers on a standard K8S environment.
+    Hashing,
+}
+
+/// A scheduler implementation that remembers where pods were once scheduled (based on
+/// their ids) and maps them to the same nodes in the future. The `history` provides preferred
+/// nodes to map onto based past mappings.
+/// The `strategy` might choose a different node if the history node cannot be used.
+pub struct StickyScheduler<'a, H: PodPlacementHistory> {
+    pub history: &'a mut H,
+    pub strategy: ScheduleStrategy,
+}
+
+//--------------------------------------------------------------------------------
+// Implementation
+//--------------------------------------------------------------------------------
+
+impl<'a> K8SUnboundedHistory<'a> {
+    pub fn new(client: &'a Client, history: PodToNodeMapping) -> Self {
+        K8SUnboundedHistory {
+            client,
+            history,
+            modified: false,
+        }
+    }
+
+    pub async fn save<T>(&mut self, resource: &T) -> OperatorResult<T>
+    where
+        T: Clone + Debug + DeserializeOwned + Resource,
+        <T as Resource>::DynamicType: Default,
+    {
+        if self.modified {
+            return match self
+                .client
+                .merge_patch_status(resource, &json!({ "history": self.history }))
+                .await
+            {
+                Ok(res) => {
+                    self.modified = false;
+                    Ok(res)
+                }
+                err => err,
+            };
+        }
+
+        Ok(resource.clone())
+    }
+}
+
+impl SchedulerState {
+    pub fn new(current_mapping: PodToNodeMapping, new_mapping: PodToNodeMapping) -> Self {
+        SchedulerState {
+            current_mapping,
+            new_mapping,
+        }
+    }
+
+    pub fn mapping(&self) -> PodToNodeMapping {
+        self.current_mapping.merge(&self.new_mapping)
+    }
+
+    pub fn new_mapping(&self) -> PodToNodeMapping {
+        self.new_mapping.clone()
+    }
+}
+
+impl TryFrom<String> for PodIdentity {
+    type Error = Error;
+    fn try_from(s: String) -> Result<Self, Error> {
+        let split = s.split(';').collect::<Vec<&str>>();
```

Comment:
We should investigate if ";" really is a good seperator and nothing can interfere.

Reply:
We decided to leave it as is. A warning is logged if any of the pod identity fields contains a semicolon. This restriction might be removed in the future if a different implementation of the pod placement history will come up.